### PR TITLE
[REVIEW] sorted_merge functionality

### LINKED ISSUE #1344: [FEA] Sorted merge
NOTE: the solution to this issue is already being implemented. The API below is the current implemented API

**Is your feature request related to a problem? Please describe.**
I need a way to take two data frames that have been already sorted and put them into a single data frame and maintain its sorted order.

**Describe the solution you'd like**

```
 @brief Merge sorted arrays of gdf_column.
 
 @Param[in] left_cols A column array to be merged with right_cols
 @Param[in] right_cols A column array to be merged with left_cols
 @Param[in] ncols Column length in left_cols and right_cols
 @Param[in] sort_by_cols Indices of left_cols and right_cols to be used
                          for comparison criteria
 @Param[in] asc_desc Sort order types of columns indexed by sort_by_cols
 @Param[out] output_cols Merged columns
 
 @Returns GDF_SUCCESS upon successful merge
 
gdf_error gdf_sorted_merge(gdf_column const * const *     left_cols,
                                                 gdf_column const * const *     right_cols,
                                                 const gdf_size_type ncols,
                                                 gdf_column const *      sort_by_cols,
                                                 gdf_column const *      asc_desc,
                                                 gdf_column **     output_cols);
```
**Describe alternatives you've considered**

**Additional context**
In particular need this for the C++ API

### PR DESCRIPTION
**Pull request by jeanp413**
Description sorted_merge
Merges two sorted tables (including null values) into one sorted table containing data from both tables.

Example

input:
table 1:
col 1 {0, 1, 2, 3}
col 2 {4, 5, 6, 7}
table 2:
col 1 {1, 2}
col 2 {8, 9}

output table:
{0, 1, 1, 2, 2, 3}
{4, 5, 8, 6, 9, 7}

closes #1344

### CODE REVIEW

**Comment by harrism** [comment]
We have two days of burn down left for 0.8, so I'm going to move this to 0.9.

File: cpp/include/sorted_merge.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ *                         for comparison criteria
+ * @Param[in] asc_desc Sort order types of columns indexed by sort_by_cols
+ * @Param[in] nulls_are_smallest Flag indicating is nulls are to be treated as the smallest value
+ *
+ * @Returns A table containing sorted data from left_table and right_table
+ */
+table sorted_merge(table const& left_table,
+                   table const& right_table,
+                   std::vector<gdf_size_type> const& sort_by_cols,
+                   rmm::device_vector<int8_t> const& asc_desc,
```
Comment (jrhemstad) [inline cpp/include/sorted_merge.hpp]: For ease of use, make this a host vector and then you can copy it to the device internally.

Also, instead of using an `int8_t`, this should use a boolean or an enum. 
```suggestion
                   std::vector<int8_t> const& asc_desc,
```

File: cpp/include/sorted_merge.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * @Param[in] sort_by_cols Indices of left_cols and right_cols to be used
+ *                         for comparison criteria
+ * @Param[in] asc_desc Sort order types of columns indexed by sort_by_cols
+ * @Param[in] nulls_are_smallest Flag indicating is nulls are to be treated as the smallest value
+ *
+ * @Returns A table containing sorted data from left_table and right_table
+ */
+table sorted_merge(table const& left_table,
+                   table const& right_table,
+                   std::vector<gdf_size_type> const& sort_by_cols,
```
Comment (jrhemstad) [inline cpp/include/sorted_merge.hpp]: A little clearer to refer to these as the "key" columns. 
```suggestion
                   std::vector<gdf_size_type> const& key_cols,
```

File: cpp/src/sortedmerge/sortedmerge.cu [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+    return std::make_pair(outputTableIndices, outputRowIndices);
+}
+
+} // namespace
+
+namespace cudf {
+namespace detail {
+
+table sorted_merge(table const& left_table,
```
Comment (harrism) [inline cpp/src/sortedmerge/sortedmerge.cu]: To make later stream support easier, this `detail::sorted_merge` should take a stream that is defaulted to zero. Then use the stream for all execution, allocation, and copies inside the function.

File: cpp/tests/sorted-merge/sorted_merge-test.cu [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    const gdf_size_type outputLength = leftColumn1->size + rightColumn1->size;
+    cudf::test::column_wrapper<TypeParam> expectedDataWrap1({0, 1, 1, 2, 2, 3}, std::vector<gdf_valid_type>(gdf_valid_allocation_size(outputLength),0xFF));
+    cudf::test::column_wrapper<TypeParam> expectedDataWrap2({4, 8, 5, 9, 6, 7}, std::vector<gdf_valid_type>(gdf_valid_allocation_size(outputLength),0xFF));
+
+    EXPECT_TRUE(gdf_equal_columns<TypeParam>(expectedDataWrap1.get(), outputTable.get_column(0)));
+    EXPECT_TRUE(gdf_equal_columns<TypeParam>(expectedDataWrap2.get(), outputTable.get_column(1)));
+}
+
+TYPED_TEST(SortedMergeTest, Merge2KeyNullColumns) {
+    cudf::test::column_wrapper<TypeParam> leftColWrap1({0, 1, 2, 3, -1}, bools_to_valids({1, 1, 1, 1, 0}));
```
Comment (harrism) [inline cpp/tests/sorted-merge/sorted_merge-test.cu]: Why not use a `column_wrapper` constructor that takes generator lambdas -- then you won't have to have `bools_to_valids`, it does it for you.

File: cpp/tests/sorted-merge/sorted_merge-test.cu [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            gdf::util::turn_bit_off(vec.data(), i);
+        }
+    }
+    return vec;
+}
+
+template <typename T>
+class SortedMergeTest : public GdfTest {};
+
+using SortedMergerTypes = ::testing::Types<int8_t, int32_t, double>;
```
Comment (harrism) [inline cpp/tests/sorted-merge/sorted_merge-test.cu]: I think you should test on all types that can be merged.

File: cpp/tests/sorted-merge/sorted_merge-test.cu [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+TYPED_TEST_CASE(SortedMergeTest, SortedMergerTypes);
+
```
Comment (harrism) [inline cpp/tests/sorted-merge/sorted_merge-test.cu]: Your test cases are all very small vectors. you could have bugs (e.g. in your CUDA kernel) that only happen when the kernel is run with more than one thread block, and I don't think any of your small tests will exercise that case.

File: cpp/tests/sorted-merge/sorted_merge-test.cu [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    EXPECT_TRUE(gdf_equal_columns<TypeParam>(expectedDataWrap2.get(), outputTable.get_column(1)));
+}
+
+
+class SortedMergeStringTest : public GdfTest {};
+
+TEST_F(SortedMergeStringTest, Merge1KeyColumns) {
+    constexpr int STRING_LENGTH = 8;
+
+    const char * leftStrings1[] = {"aaaaaaab", "aaaaaaaf", "aaaaaaak", "aaaaaaap", "aaaaaaat", "aaaaaaax", "aaaaaaaz"};
```
Comment (harrism) [inline cpp/tests/sorted-merge/sorted_merge-test.cu]: Consider just stringifying your integer input data so that you can use the exact same tests for all data types. See copy_range_tests.cu and stream_compaction_tests.cu for examples.

File: cpp/tests/sorted-merge/sorted_merge-test.cu [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        } else {
+            gdf::util::turn_bit_off(vec.data(), i);
+        }
+    }
+    return vec;
+}
+
+template <typename T>
+class SortedMergeTest : public GdfTest {};
+
```
Comment (harrism) [inline cpp/tests/sorted-merge/sorted_merge-test.cu]: Need a test for error cases -- e.g. mismatched column dtypes. Basically anything you have a CUDF_EXPECTS for.

File: cpp/include/sorted_merge.hpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+ * @Param[in] left_table A sorted table to be merged
+ * @Param[in] right_table A sorted table to be merged
+ * @Param[in] sort_by_cols Indices of left_cols and right_cols to be used
+ *                         for comparison criteria
+ * @Param[in] asc_desc Sort order types of columns indexed by sort_by_cols
+ * @Param[in] nulls_are_smallest Flag indicating is nulls are to be treated as the smallest value
+ *
+ * @Returns A table containing sorted data from left_table and right_table
+ */
+table sorted_merge(table const& left_table,
```
Comment (harrism) [inline cpp/include/sorted_merge.hpp]: I wonder if this should be called `ordered_merge` or just `merge`. I'm just wary of confusion over whether the function performs a sort. Thrust just names it `merge`, as does moderngpu.

File: cpp/src/sortedmerge/sortedmerge.cu [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+  gdf_index_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+
+  auto active_threads =
+      __ballot_sync(0xffffffff, destination_row < num_destination_rows);
+
+  while (destination_row < num_destination_rows) {
+    gdf_valid_type const* source_mask = merged_table_indices[destination_row] == LEFT_TABLE ? source_left_mask : source_right_mask;
+    bool const source_bit_is_valid{
+        gdf_is_valid(source_mask, merged_row_indices[destination_row])};
```
[the remark this answers is not in the document]
Comment (jeanp413) [reply cpp/src/sortedmerge/sortedmerge.cu]: The kernel receives two bit masks so there would be actually four cases. No sure if it's OK to specialize for all of them. Or, maybe, make the kernel just accept one bit mask and launch it two times?

File: cpp/tests/merge/merge-test.cu [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+#include "tests/utilities/column_wrapper_factory.hpp"
+#include "tests/utilities/cudf_test_fixtures.h"
+#include "tests/utilities/nvcategory_utils.cuh"
+
+template <typename T>
+class MergeTest : public GdfTest {};
+
+using test_types =
+  ::testing::Types<int8_t, int16_t, int32_t, int64_t, float, double,
+                   /* cudf::bool8, */ cudf::nvstring_category>;
```
Comment (harrism) [inline cpp/tests/merge/merge-test.cu]: Why is cudf::bool8 commented out?

Comment (harrism) [reply cpp/tests/merge/merge-test.cu]: Am I correct that you still aren't testing anything with rows bigger than a single thread block can handle?

Comment (jeanp413) [reply cpp/tests/merge/merge-test.cu]: I commented it out because of the conversion `0 => false any other value => true` doesn't map nicely to the labmda functions given to the `expectedData` column_wrappers. I could specific unit test for `cudf::bool8` if needed.
Already pushed a3e517174ed83a94014452d0e12b2a64751bf41a which updates the size of the test data.

Comment (harrism) [reply cpp/tests/merge/merge-test.cu]: I think an easy change would be to check the source data column's dtype and if it is GDF_BOOL8, then change the lambda for the input data and the ExpectedData* to do something more interesting for bools, like return `row % 2` instead of `row`.

Comment (jeanp413) [reply cpp/tests/merge/merge-test.cu]: Updated the unit test for `cudf::bool8` though they are not that interesting because the input data must be sorted so its just a bunch of 0's followed by 1's or vice versa

Comment (harrism) [reply cpp/tests/merge/merge-test.cu]: That's OK. The point is to exercise that pathway and ensure there are no surprises.

File: cpp/include/sorted_merge.hpp (hunk 9 above)
Comment (jrhemstad) [reply cpp/include/sorted_merge.hpp]: Could possibly be confusing with Pandas `merge` which is really a `join` :upside_down_face:

File: cpp/include/cudf/merge.hpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+ *                     for comparison criteria
+ * @Param[in] asc_desc Sort order types of columns indexed by key_cols
+ * @Param[in] nulls_are_smallest Flag indicating is nulls are to be treated as the smallest value
+ *
+ * @Returns A table containing sorted data from left_table and right_table
+ */
+table merge(table const& left_table,
+            table const& right_table,
+            std::vector<gdf_size_type> const& key_cols,
+            std::vector<order_by_type> const& asc_desc,
```
Comment (jrhemstad) [inline cpp/include/cudf/merge.hpp]: Instead of order_by_type, can this just be a bool? Is there any other option besides ascending or descending?

File: cpp/src/merge/merge.cu [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+    CUDF_EXPECTS(key_cols.size() == asc_desc.size(), "Mismatched size between key_cols and asc_desc");
+
+
+    auto gdf_col_deleter = [](gdf_column *col) {
+        gdf_column_free(col);
+        delete col;
+    };
+    using gdf_col_ptr = typename std::unique_ptr<gdf_column, decltype(gdf_col_deleter)>;
+    std::vector<gdf_col_ptr> temp_columns_to_free;
+    std::vector<gdf_column*> left_cols_sync(const_cast<gdf_column**>(left_table.begin()), const_cast<gdf_column**>(left_table.end()));
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: Why is the `const`ness of the input being casted away? That is rarely a good idea.

File: cpp/src/merge/merge.cu [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+    std::vector<gdf_column*> right_cols_sync(const_cast<gdf_column**>(right_table.begin()), const_cast<gdf_column**>(right_table.end()));
+    for (gdf_size_type i = 0; i < left_table.num_columns(); i++) {
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: This is the kind of functionality that should be moved into a utility function. Luckily, it already exists: https://github.com/rapidsai/cudf/blob/branch-0.9/cpp/src/string/nvcategory_util.hpp#L45

File: cpp/src/merge/merge.cu [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+                                                    merged_row_indices.begin()));
+    auto index_end_it = thrust::make_zip_iterator(thrust::make_tuple(
+                                                thrust::make_counting_iterator(static_cast<gdf_size_type>(merged_table_indices.size())),
+                                                merged_table_indices.end(),
+                                                merged_row_indices.end()));
+
+    thrust::for_each(rmm::exec_policy(stream)->on(stream),
+                    index_start_it,
+                    index_end_it,
+                    [=] __device__ (auto const & idx_tuple){
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: Passing tuples as input parameters can be tricky to read and understand. It helps to break up a tuple into named variables. For example:
```
[=] __device__ (auto const & idx_tuple){
   thing0 = thrust::get<0>(idx_tuple);
   thing1 = thrust::get<1>(idx_tuple);
   thing2 = thrust::get<2>(idx_tuple);
```

File: cpp/src/merge/merge.cu [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+                        device_table const& right_table,
+                        rmm::device_vector<int8_t> const& asc_desc,
+                        bool nulls_are_smallest,
+                        cudaStream_t stream) {
+
+    const gdf_size_type left_size  = left_table.num_rows();
+    const gdf_size_type right_size = right_table.num_rows();
+    const gdf_size_type total_size = left_size + right_size;
+
+    auto left_side = thrust::make_constant_iterator(static_cast<gdf_size_type>(LEFT_TABLE));
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: I think this is a case where using the `make_*_iterator` is actually more verbose than just using the type directly:
```suggestion
    thrust::constant_iterator<gdf_size_type> left_side{LEFT_SIDE};;
```

File: cpp/src/merge/merge.cu [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+#include "utilities/cuda_utils.hpp"
+
+namespace {
+
+using bit_mask::bit_mask_t;
+
+/**
+ * @brief Source table identifier to copy data from.
+ */
+enum side_value { LEFT_TABLE = 0, RIGHT_TABLE };
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: This is a good place to use an `enum class` backed by a `bool`. Since you only need to store 2 values, make that explicit by using a `bool`. 
```suggestion
enum class side : bool { LEFT, RIGHT };
```

File: cpp/src/merge/merge.cu [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+    auto left_indices = thrust::make_counting_iterator(static_cast<gdf_size_type>(0));
+    auto right_indices = thrust::make_counting_iterator(static_cast<gdf_size_type>(0));
+
+    auto left_begin_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(left_side, left_indices));
+    auto right_begin_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(right_side, right_indices));
+
+    auto left_end_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(left_side + left_size, left_indices + left_size));
+    auto right_end_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(right_side + right_size, right_indices + right_size));
+
+    rmm::device_vector<gdf_size_type> out_table_indices(total_size);
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: It's inefficient to use `gdf_size_type` (which is a 4B type) to store what can only be one of two values. 

Use the `enum` as the type instead to make it clearer what this is storing. And since the enum is changed to be backed by a `bool`, it will consume 1/4 the memory.

File: cpp/src/merge/merge.cu [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+    rmm::device_vector<gdf_size_type> out_row_indices(total_size);
+    auto output_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(out_table_indices.begin(), out_row_indices.begin()));
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: This would be simplified by instead using a vector of `thrust::pair` instead of two independent vectors.

```
using index_type = thrust::pair< side, gdf_size_type > ; // `first` indices left/right side, `second` indicates the row index
thrust::device_vector<index_type> merged_indices(total_size);
```

File: cpp/src/merge/merge.cu [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+    std::transform(key_cols.cbegin(), key_cols.cend(), right_key_cols_vect.begin(),
+                  [&] (gdf_index_type const index) { return right_sync_table.get_column(index); });
+
+    auto left_key_table = device_table::create(left_key_cols_vect.size(), left_key_cols_vect.data());
+    auto right_key_table = device_table::create(right_key_cols_vect.size(), right_key_cols_vect.data());
+    rmm::device_vector<int8_t> asc_desc_d(asc_desc);
+
+    rmm::device_vector<gdf_size_type> merged_table_indices;
+    rmm::device_vector<gdf_size_type> merged_row_indices;
+    std::tie(merged_table_indices, merged_row_indices) = generate_merged_indices(*left_key_table, *right_key_table, asc_desc_d, nulls_are_smallest, stream);
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: Using a vector of `thrust::pair` would simplify this. See above.

File: cpp/src/merge/merge.cu [hunk 21]
```diff
@@ -0,0 +1,329 @@
...
+
+void materialize_bitmask(gdf_column const* left_col,
+                        gdf_column const* right_col,
+                        gdf_column* out_col,
+                        gdf_size_type const* table_indices,
+                        gdf_size_type const* row_indices,
+                        cudaStream_t stream) {
+    constexpr gdf_size_type BLOCK_SIZE{256};
+    cudf::util::cuda::grid_config_1d grid_config {out_col->size, BLOCK_SIZE };
+    // TODO: cant use just one template flag for valids, so need to test for null explicitly
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: Why not? Since you're invoking a kernel per column, you should be able to just test each column for the presence of null values and then invoke the proper instantiation.

Comment (jeanp413) [reply cpp/src/merge/merge.cu]: The kernel receives two bit masks so there would be actually four cases. No sure if it's OK to specialize for all of them. Or make the kernel just accept one bit mask and launch it two times?

Comment (jrhemstad) [reply cpp/src/merge/merge.cu]: I'd just use two template bool parameters. I very much doubt the compile time will suffer significantly. For example, see what I did here for groupby: https://github.com/rapidsai/cudf/blob/branch-0.9/cpp/src/groupby/hash/groupby.cu#L408

File: cpp/src/merge/merge.cu (hunk 14 above)
Comment (jeanp413) [reply cpp/src/merge/merge.cu]: I'm using `sync_column_categories` [here](https://github.com/rapidsai/cudf/pull/2049/files#diff-ef8fa6376fd285383b7c5011ef5af737R237). Did you mean something else?

Comment (jrhemstad) [reply cpp/src/merge/merge.cu]: Ah, I see, you're right. I guess it just got lost in all the extra code here that is necessary in order to call it.

File: cpp/src/merge/merge.cu (hunk 13 above)
Comment (jeanp413) [reply cpp/src/merge/merge.cu]: Because the `cudf::table` constructor needs `gdf_column**` or  `std::vector<gdf_column*>`.
`table.begin()` returns `gdf_column const* const*` but the underlying data is `std::vector<gdf_column*>` so it's safe to do the cast. Maybe `table.begin()` should always return `gdf_column**`?

File: cpp/include/cudf/merge.hpp (hunk 12 above)
Comment (jeanp413) [reply cpp/include/cudf/merge.hpp]: I think it would nice to be consistent with the `order_by_type` enum type (this could be `enum class order_by_type : bool` in the future?) besides this is just in cpu memory, later, I [convert it](https://github.com/rapidsai/cudf/pull/2049/files#diff-ef8fa6376fd285383b7c5011ef5af737R256) to a `rmm::device_vector<int8_t>`

File: cpp/src/merge/merge.cu (hunk 19 above)
Comment (jeanp413) [reply cpp/src/merge/merge.cu]: For merge the input should be convertible to output, I tried `thrust::make_zip_iterator(thrust::make_pair(left_side, left_indices))` as input but it doesn't work. Do you know how to initialize the input?

Comment (jrhemstad) [reply cpp/src/merge/merge.cu]: I don't think you need a zip iterator anymore, because you're not zipping together two independent iterators. Instead you've got a single iterator iterating across a range of `thrust::pair`.

File: cpp/include/cudf/merge.hpp (hunk 12 above)
Comment (williamBlazing) [reply cpp/include/cudf/merge.hpp]: Since internally it uses a rmm::device_vector<int8_t>, then i dont think we should have it as a bool. I know it makes sense to have it as a bool, but I think that as an enum its also nicely descriptive

Comment (jrhemstad) [reply cpp/include/cudf/merge.hpp]: The change instead should then be to not use an `int8_t` internally.

File: cpp/src/merge/merge.cu [hunk 22]
```diff
@@ -0,0 +1,348 @@
...
+        CUDF_TRY( sync_column_categories(tmp_arr_input, tmp_arr_output, 2) );
+
+        left_cols_sync[i] = new_left_column_ptr;
+        right_cols_sync[i] = new_right_column_ptr;
+    }
+
+    table left_sync_table(left_cols_sync.data(), left_cols_sync.size());
+    table right_sync_table(right_cols_sync.data(), right_cols_sync.size());
+
+    std::vector<gdf_column*> left_key_cols_vect(key_cols.size());
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: It looks like the only reason you're creating `left_key_cols_vect` and `right_key_cols_vect` is to pass into `device_table::create`, but that's unneccessary, because you can pass the `table` directly into `device_table::create`. 

https://github.com/rapidsai/cudf/blob/branch-0.9/cpp/src/table/device_table.cuh#L94

File: cpp/src/merge/merge.cu [hunk 23]
```diff
@@ -0,0 +1,348 @@
...
+                        side src_side = thrust::get<0>(merged_idx);
+                        gdf_size_type src_row  = thrust::get<1>(merged_idx);
+                        device_table const & src_device_table = src_side == side::LEFT ? left_device_table : right_device_table;
+                        copy_row<false>(output_device_table, dest_row, src_device_table, src_row);
+                    });
+    
+    CHECK_STREAM(0);
+
+    if (nullable) {
+        for (gdf_size_type i = 0; i < destination_table.num_columns(); i++) {
```
Comment (jrhemstad) [inline cpp/src/merge/merge.cu]: This could be nicely implemented as a `std::transform` where the inputs are `left_sync_table` and `right_sync_table`, and the output is `destination_table`. `table::begin` and `table::end` are provided for exactly this kind of thing.

File: cpp/include/cudf/merge.hpp (hunk 12 above)
Comment (williamBlazing) [reply cpp/include/cudf/merge.hpp]: This would involve changing row_inequality_comparator and its usage by orderby as well. That is why we were not considering that

Comment (jrhemstad) [reply cpp/include/cudf/merge.hpp]: Yeah, that's fine. Updating all of that can happen in a single, follow on PR.

File: cpp/src/merge/merge.cu (hunk 22 above)
Comment (jeanp413) [reply cpp/src/merge/merge.cu]: I need them because I'm using just the `key_cols` columns of the `left_sync_table` and `right_sync_table` not the whole table

File: cpp/src/merge/merge.cu (hunk 23 above)
Comment (jeanp413) [reply cpp/src/merge/merge.cu]: I'm not following. Do you mean something like this?
```c++
std::transform(left_sync_table.begin(),
               left_sync_table.end(),
               right_sync_table.begin(),
               destination_table.begin(),
               [&, =stream](gdf_column const* left_col, gdf_column const* right_col){
                   /* destination_table already has the merged data, 
                    * need the same gdf_column to update the valids
                    * and null_count and return it */
                   int index = /*calculate which column index*/;
                   gdf_column* out_col = destination_table.get_column(index);
                   materialize_bitmask(left_col, right_col, out_col, merged_indices.data().get(), stream);
                   out_col->null_count = left_col->null_count + right_col->null_count;
                   return out_col;
               });
```

Comment (jrhemstad) [reply cpp/src/merge/merge.cu]: Ah, I see the problem is you need to modify `out_col` in place. Nevermind.

File: cpp/tests/CMakeLists.txt [hunk 24]
```diff
@@ -348,6 +348,15 @@ set(DEVICE_TABLE_TEST_SRC
 
 ConfigureTest(DEVICE_TABLE_TEST "${DEVICE_TABLE_TEST_SRC}")
 
+###################################################################################################
+# - sorted-merge tests -------------------------------------------------------------------------------------
+
+set(MERGE_TEST_SRC
+    "${CMAKE_CURRENT_SOURCE_DIR}/merge/merge-test.cu"
+    "${CMAKE_CURRENT_SOURCE_DIR}/utilities/nvcategory_utils.cu")
```
Comment (jrhemstad) [inline cpp/tests/CMakeLists.txt]: This shouldn't be necessary because `nvcategory_utils.cu` is already linked into `cudftestutil` lib. 

```suggestion
    "${CMAKE_CURRENT_SOURCE_DIR}/utilities/nvcategory_utils.cu")
```

File: cpp/tests/merge/merge-test.cu [hunk 25]
```diff
@@ -0,0 +1,466 @@
...
+    gdf_column *rightColumn2 = rightColWrap2.get();
+
+    gdf_column *leftColumns[]  = {leftColumn1, leftColumn2};
+    gdf_column *rightColumns[] = {rightColumn1, rightColumn2};
+
+    std::vector<gdf_size_type> sortByCols = {0, 1};
+    std::vector<order_by_type> orderByTypes = {GDF_ORDER_ASC, GDF_ORDER_DESC};
+
+    cudf::table outputTable;
+    EXPECT_NO_THROW(outputTable = cudf::merge(cudf::table(leftColumns, 2),
```
Comment (jrhemstad) [inline cpp/tests/merge/merge-test.cu]: Just FYI, `cudf::table` has a `std::initializer_list<gdf_column*>` constructor that can simplify your code here a bit. 
This way you don't need to make an explicit array of `gdf_column*` in `leftColumns[]` and `rightColumns[]` 

```suggestion
    EXPECT_NO_THROW(outputTable = cudf::merge(cudf::table({leftColWrap1.get(), leftColWrap2.get()}),
```

**Comment by jrhemstad** [comment]
rerun tests

**Comment by jrhemstad** [comment]
@jeanp413 once this PR is complete, it would be great if you could also tackle https://github.com/rapidsai/cudf/issues/2312

File: cpp/src/copying/slice.cu [hunk 26]
```diff
@@ -215,9 +215,7 @@ public:
       }
 
       if (output_column->dtype == GDF_STRING_CATEGORY){
-        NVCategory* new_category = static_cast<NVCategory*>(input_column_.dtype_info.category)->gather_and_remap(
-                      static_cast<int*>(output_column->data), (unsigned int)output_column->size);
-        output_column->dtype_info.category = new_category;
+        CUDF_TRY(nvcategory_gather(output_column, static_cast<NVCategory*>(input_column_.dtype_info.category)));
```
Comment (jrhemstad) [inline cpp/src/copying/slice.cu]: Why is `slice` being changed in this PR?

Comment (jeanp413) [reply cpp/src/copying/slice.cu]: Because I updated nvcategory_utils.cu 
```diff
-    NVStrings* tptr = static_cast<NVCategory*>(column->dtype_info.category)->to_strings();
+    NVStrings* tptr = static_cast<NVCategory*>(column->dtype_info.category)->gather_strings(static_cast<nv_category_index_type*>(column->data),
+                                                                                            column->size,
+                                                                                            DEVICE_ALLOCATED);
```
with `to_strings` I was getting the wrong order, but slice unit test started crashing, changing to `nvcategory_gather` fixed that which also, I think, is the correct function instead of `gather_and_remap`
